### PR TITLE
Migration to Debian 12

### PR DESCRIPTION
**Pull request by fgalan**
- [x] Code changes
- [x] Check unit test in Debian 12 locally
- [x] Check ftest in Debian 12 locally (except the usual 3001_mqtt/mqtt_subscription_multibroker.test glith)
- [x] Dockerfile (test build locally) (`docker build -t prueba --build-arg GIT_REV_ORION=hardening/debian12-migration .`) - [224ace3](https://github.com/telefonicaid/fiware-orion/pull/4417/commits/224ace3e33c8f26a355146a66d9b6229c98a2eba)
- [x] Check valgrind run in Debian 12 locally - detail [here](https://github.com/telefonicaid/fiware-orion/pull/4417#issuecomment-1705528897)
- [x] Check coverage in Debian 12 locally - problems found doing this, but not related with this PR istelf (moved to issue #4418), good enough with [this result](https://github.com/telefonicaid/fiware-orion/pull/4417#issuecomment-1706149540)
- [x] CI Dockerfile (test build locally) - [2d870f7](https://github.com/telefonicaid/fiware-orion/pull/4417/commits/2d870f70d8866b7f22291c23362d74a6c5b9cb00)
- [x] Test after rebuilding CI with the changes in CI Dockerfile (use manual build on dockerhub) (*) - at [c3c827d](https://github.com/telefonicaid/fiware-orion/pull/4417/commits/c3c827d3103e44a8f2ff5acd51a79f03e4302b2f) tests are passing fine
- [x] Documentation
- [x] CNR
- [ ] (post merge) relaunch tests in PR https://github.com/telefonicaid/fiware-orion/pull/4331, PR https://github.com/telefonicaid/fiware-orion/pull/4332 and PR https://github.com/telefonicaid/fiware-orion/pull/4414 to check regression in Debian 12

(*) Up to this point, regression in this PR commits is done using preexisting Debian 11 based CI image

### CODE REVIEW

**Comment by fgalan** [comment]
At commit [b65186d](https://github.com/telefonicaid/fiware-orion/pull/4417/commits/b65186dad61c8ae994b9640cc5bd70b3cca03fd7), valgrind pass seems ok:

```
test/valgrind/valgrindTestSuite.sh
lun 04 sep 2023 18:37:35 CEST
Run tests 0 to 0
Test 001/1133: 0000_bad_requests/exit ............................................................................................................... OK (.07 seconds)
Test 002/1133: 0000_bad_requests/pagination_error_in_uri_params ..................................................................................... OK (11.04 seconds)
Test 003/1133: 0000_cli/bool_option_with_value ...................................................................................................... Not apt for valgrind - skipped
Test 004/1133: 0000_cli/command_line_options ........................................................................................................ Not apt for valgrind - skipped
Test 005/1133: 0000_cli/tracelevel_without_logLevel_as_DEBUG ........................................................................................ Not apt for valgrind - skipped
Test 006/1133: 0000_cli/version ..................................................................................................................... Not apt for valgrind - skipped
Test 007/1133: 0000_content_related_headers/accept_fail_01 .......................................................................................... OK (8.89 seconds)
Test 008/1133: 0000_content_related_headers/accept_fail ............................................................................................. OK (12.06 seconds)
Test 009/1133: 0000_content_related_headers/missing_content_type_header ............................................................................. OK (10.55 seconds)
Test 010/1133: 0000_deprecated_checkings/disable_ngsi_cli ........................................................................................... OK (9.47 seconds)
Test 011/1133: 0000_deprecated_checkings/geoquery_circle_deprecated ................................................................................. OK (15.94 seconds)
Test 012/1133: 0000_deprecated_checkings/get_entity_dates_with_options .............................................................................. OK (47.74 seconds) (no leak but ftest error 1)
Test 013/1133: 0000_deprecated_checkings/legacy_geoquery_bad_coords ................................................................................. OK (12.05 seconds)
Test 014/1133: 0000_deprecated_checkings/legacy_geoquery_circle_json ................................................................................ OK (16.65 seconds)
Test 015/1133: 0000_deprecated_checkings/legacy_geoquery_polygon_json ............................................................................... OK (17.42 seconds)
Test 016/1133: 0000_deprecated_checkings/legacy_location_no_actual_location_change .................................................................. OK (11.07 seconds)
Test 017/1133: 0000_deprecated_checkings/legacy_wgs84 ............................................................................................... OK (15.70 seconds)
Test 018/1133: 0000_deprecated_checkings/log_deprecate_warning_dynamic .............................................................................. OK (10.67 seconds)
Test 019/1133: 0000_deprecated_checkings/log_deprecate_warning ...................................................................................... OK (11.65 seconds)
Test 020/1133: 0000_deprecated_checkings/metadata_id_as_regular_metadata ............................................................................ OK (12.06 seconds)
Test 021/1133: 0000_deprecated_checkings/metadata_id_duplicate_error ................................................................................ OK (11.11 seconds)
Test 022/1133: 0000_deprecated_checkings/uppercase_action_types_in_ngsiv2_batch_update .............................................................. OK (13.17 seconds)
Test 023/1133: 0000_deprecated_checkings/xml_support ................................................................................................ OK (8.75 seconds)
Test 024/1133: 0000_ipv6_support/ipv4_only .......................................................................................................... OK (.05 seconds)
Test 025/1133: 0000_ipv6_support/ipv6_only .......................................................................................................... OK (.07 seconds)
Test 026/1133: 0000_json_parse/jsonParseGetNotifyContext ............................................................................................ OK (9.15 seconds)
Test 027/1133: 0000_json_parse/jsonParsePostQueryContext ............................................................................................ OK (11.23 seconds)
Test 028/1133: 0000_large_requests/large_query_context .............................................................................................. OK (10.81 seconds)
Test 029/1133: 0000_large_requests/large_update_context ............................................................................................. OK (9.18 seconds)
...
```

A full pass has not been done, given it takes a long time, it will be done by GitActions once this PR gets merged.

**Comment by fgalan** [comment]
```
$ make coverage_unit_test
...
Overall coverage rate:
  lines......: 21.4% (4888 of 22891 lines)
  functions..: 41.9% (753 of 1799 functions)
```

The actual results (pretty low) is not relevant (this is due we are no relying in unit test too much since MongoDB driver changed time ago in Orion 3.0.0). What is relevant is that the coverage tooling works.

File: test/functionalTest/testHarness.sh [hunk 1]
```diff
@@ -914,7 +914,8 @@ function testDisabled
   dIx=0
   while [ $dIx -lt  ${#DISABLED[@]} ]
   do
-    if [ test/functionalTest/cases/$testcase == ${DISABLED[$dIx]} ]
+    # Comparison is done based in filename, skipping the path (see https://stackdiary.com/tutorials/bash-get-filename-from-path/)
+    if [ "${testcase##*/}" == "${DISABLED[$dIx]##*/}" ]
```
Comment (fgalan) [inline test/functionalTest/testHarness.sh:918]: Not actually related with this PR, but found while I was investigating make coverage issues.

NTC (informative)

**Comment by fgalan** [comment]
At commit [2d870f7](https://github.com/telefonicaid/fiware-orion/pull/4417/commits/2d870f70d8866b7f22291c23362d74a6c5b9cb00) first run is with preexisting Debian 11 image

imagen

imagen

Let's build now the new Debian 12 based image in dockerhub and see how it goes...

**Comment by fgalan** [comment]
Fails in cases/1705_csub_cache_objects/csub_cache_objects_errors.test due to curl CLI has changed:

Debian 11:

```
$ curl 'http://localhost:9999/v2/entities?q=a==12;    '
{"error":"BadRequest","description":"empty q-item"}

$ curl --version
curl 7.74.0 (x86_64-pc-linux-gnu) libcurl/7.74.0 OpenSSL/1.1.1n zlib/1.2.11 brotli/1.0.9 libidn2/2.3.0 libpsl/0.21.0 (+libidn2/2.3.0) libssh2/1.9.0 nghttp2/1.43.0 librtmp/2.3
Release-Date: 2020-12-09
Protocols: dict file ftp ftps gopher http https imap imaps ldap ldaps mqtt pop3 pop3s rtmp rtsp scp sftp smb smbs smtp smtps telnet tftp 
Features: alt-svc AsynchDNS brotli GSS-API HTTP2 HTTPS-proxy IDN IPv6 Kerberos Largefile libz NTLM NTLM_WB PSL SPNEGO SSL TLS-SRP UnixSockets
```

Debian 12:

```
$ curl 'http://localhost:9999/v2/entities?q=a==12;    '
curl: (3) URL using bad/illegal format or missing URL

$ curl --version
curl 7.88.1 (x86_64-pc-linux-gnu) libcurl/7.88.1 OpenSSL/3.0.9 zlib/1.2.13 brotli/1.0.9 zstd/1.5.4 libidn2/2.3.3 libpsl/0.21.2 (+libidn2/2.3.3) libssh2/1.10.0 nghttp2/1.52.0 librtmp/2.3
Release-Date: 2023-02-20
Protocols: dict file ftp ftps gopher gophers http https imap imaps ldap ldaps mqtt pop3 pop3s rtmp rtsp scp sftp smb smbs smtp smtps telnet tftp
Features: alt-svc AsynchDNS brotli GSS-API HSTS HTTP2 HTTPS-proxy IDN IPv6 Kerberos Largefile libz NTLM NTLM_WB PSL SPNEGO SSL threadsafe TLS-SRP UnixSockets zstd
```

Thus, it seem curl has changed its behaviour among versions...

I have posted to SOF about this: https://stackoverflow.com/questions/77045755/avoid-curl-to-check-for-malformed-urls

**Comment by fgalan** [comment]
@fisuda, several .md files in the English version have been modified in this PR. I have partially modified some Japanese .md files to align, but only partially (mainly numeric information). It would be great if you could have a look and complete this synchronization work in a PR using hardening/debian12-migration (this PR's branch) as base branch, please.

With regards to the "Ubuntu 20.04 TLS" sections, it would be great if you could check if the procedure described there is still working after the changes of this PR but that could be of course accomplished in a separate PR (related with what [discussed in this other PR](https://github.com/telefonicaid/fiware-orion/pull/4415#issuecomment-1702683613).

Thanks in advance!

**Comment by fisuda** [comment]
I sent the PR #4419.

In the current version (before merging this PR), it seems that the Orion can be built with mongo driver 1.24.3 on Ubuntu 20.04.

As of now, Ubuntu LTS (Long Term Support) version is 22.04. So I'm trying to build Orion on Ubuntu 22.04. But I enounted same error as building on debian 12. Because It seems that the base distribution of Ubuntu 22.04.3 LTS is debian 12. Atfer this PR is merged, I'd like to try to build Orion again. And I will update the "Ubuntu 20.04 LTS" sections to the "Ubuntu 22.04 LTS" sections.

```
$ cat /etc/os-release
PRETTY_NAME="Ubuntu 22.04.3 LTS"
NAME="Ubuntu"
VERSION_ID="22.04"
VERSION="22.04.3 LTS (Jammy Jellyfish)"
VERSION_CODENAME=jammy
ID=ubuntu
ID_LIKE=debian
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
UBUNTU_CODENAME=jammy
```

```
$ cat /etc/debian_version
bookworm/sid
```

Thanks

**Comment by fgalan** [comment]
> As of now, Ubuntu LTS (Long Term Support) version is 22.04. So I'm trying to build Orion on Ubuntu 22.04. But I enounted same error as building on debian 12. Because It seems that the base distribution of Ubuntu 22.04.3 LTS is debian 12. Atfer this PR is merged, I'd like to try to build Orion again. And I will update the "Ubuntu 20.04 LTS" sections to the "Ubuntu 22.04 LTS" sections.

Sounds nice! :)

File: doc/manuals/admin/build_source.md [hunk 2]
```diff
@@ -93,38 +93,43 @@ The Orion Context Broker comes with a suite of unit, valgrind and end-to-end tes
 
 * Install Google Test/Mock from sources. Previously the URL was http://googlemock.googlecode.com/files/gmock-1.5.0.tar.bz2 but Google removed that package in late August 2016 and it is no longer working.
 
-        sudo apt-get install python2
         wget https://nexus.lab.fiware.org/repository/raw/public/storage/gmock-1.5.0.tar.bz2
         tar xfvj gmock-1.5.0.tar.bz2
         cd gmock-1.5.0
         ./configure
-        sed -i 's/env python/env python2/' gtest/scripts/fuse_gtest_files.py  # little hack to make installation to work on Debian 11
+        # Adjust /path/to/fiware-orion in the next line accordingly to where you local copy of fiware-orion repo is in your system
```
Comment (mapedraza) [inline doc/manuals/admin/build_source.md:100]: Probably describing what this patch do / why is needed is a good idea

Comment (fgalan) [reply doc/manuals/admin/build_source.md:100]: It's done in the .patch file itself:

> This patch adapt the original fuse_gest_files.py to Python 3
> Check build_source.md file to see how it is used

**Review by mapedraza — APPROVED** [review]
LGTM

File: doc/manuals/admin/build_source.md (hunk 2 above)
Comment (mapedraza) [reply doc/manuals/admin/build_source.md:100]: OK, NTC

**Pull request merged**